File: test_gmaps.py
import requests
import json
from bs4 import BeautifulSoup

def extract_schema_address(url):
    print(f"Resolving: {url}")
    session = requests.Session()
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    
    try:
        res = session.get(url, headers=headers, allow_redirects=True, timeout=10)
        soup = BeautifulSoup(res.text, 'html.parser')
        
        print("\nSearching for LD+JSON Schema...")
        # Google Maps sometimes embeds standard LocalBusiness schema
        scripts = soup.find_all('script', type='application/ld+json')
        for script in scripts:
            try:
                data = json.loads(script.string)
                if '@type' in data and data['@type'] in ['LocalBusiness', 'Organization', 'Place']:
                    print(json.dumps(data.get('address', {}), indent=2))
                    return data.get('address')
            except Exception as e:
                pass
                
        print("No LD+JSON schema found.")
        
        # Another approach: find the exact address text natively rendered in the DOM
        # Often div or span containing the address has a specific class, but it varies.
        # Just rip all text and look for the zip code 18102
        print("\nScanning raw document text for the ZIP code...")
        all_text = soup.get_text(separator=' | ')
        import re
        matches = re.findall(r'[^|]*18102[^|]*', all_text)
        if matches:
            print("Found potential address snippets in text:")
            for m in set(matches):
                print(f"-> {m.strip()}")
                
    except Exception as e:
        print(f"Failed: {e}")

if __name__ == "__main__":
    extract_schema_address("https://maps.app.goo.gl/EEKjPP9oKbUENTkR8")

File: test_gmaps.js
const puppeteer = require('puppeteer');

(async () => {
    const url = 'https://maps.app.goo.gl/EEKjPP9oKbUENTkR8';
    console.log(`Launching Puppeteer to resolve: ${url}`);

    // Launch a headless browser
    const browser = await puppeteer.launch({ headless: "new" });
    const page = await browser.newPage();

    // Set a normal user agent
    await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');

    // Go to the URL and wait for the network to be idle (so React can render the DOM)
    console.log("Navigating and waiting for DOM...");
    await page.goto(url, { waitUntil: 'networkidle2' });

    // Google Maps usually stores the address in an element with a specific aria-label
    // Or we can just look for the element containing the city and state
    console.log("Evaluating the DOM...");
    const address = await page.evaluate(() => {
        // Try finding the button that contains the address text.
        // Google Maps address buttons almost always have a data-item-id starting with "address"
        const addressBtn = document.querySelector('button[data-item-id="address"]');
        if (addressBtn) {
            return addressBtn.getAttribute('aria-label') || addressBtn.innerText;
        }

        // Fallback: search all divs for a characteristic string pattern like ", ST ZIP"
        const allDivs = Array.from(document.querySelectorAll('div'));
        for (let div of allDivs) {
            if (div.innerText && div.innerText.match(/,\s*[A-Z]{2}\s*\d{5}/)) {
                return div.innerText;
            }
        }
        return "Could not find address in DOM.";
    });

    console.log(`\nAddress Found via Puppeteer:\n${address}`);

    await browser.close();
})();
